flash/core/finetuning.py

```diff
@@ -11,147 +11,169 @@
+        train_bn: bool = True,
+    ):
         """
         Args:
+            strategy_key: The finetuning strategy to be used. (Check :meth"`Trainer.finetune` for the available
+             strategies.)
+            strategy_metadata: Data that accompanies certain finetuning strategies like epoch number or number of
+             layers.
             attr_names: Name(s) of the module attributes of the model to be frozen.
             train_bn: Whether to train Batch Norm layer
         """
         super().__init__()
 
-        self.attr_names = [attr_names] if isinstance(attr_names, str) else attr_names
-        self.train_bn = train_bn
-
-    def freeze_before_training(self, pl_module: LightningModule) -> None:
-        self.freeze_using_attr_names(pl_module, self.attr_names, train_bn=self.train_bn)
-
-    def freeze_using_attr_names(self, pl_module, attr_names: List[str], train_bn: bool = True):
-        for attr_name in attr_names:
-            attr = getattr(pl_module, attr_name, None)
-            if not attr or not isinstance(attr, nn.Module):
-                MisconfigurationException(f"Your model must have a {attr} attribute")
-            self.freeze(modules=attr, train_bn=train_bn)
-
-    def finetune_function(self, pl_module: LightningModule, epoch: int, optimizer: Optimizer, opt_idx: int):
-        pass
-
-
-class Freeze(FlashBaseFinetuning):
-    def finetune_function(
+        self.strategy: FinetuningStrategies = strategy_key
+        self.strategy_metadata: Optional[Union[int, Tuple[int, int]]] = strategy_metadata
+        self.train_bn: bool = train_bn
+
+    def freeze_before_training(self, pl_module: Union[Module, Iterable[Union[Module, Iterable]]]) -> None:
+        if self.strategy != FinetuningStrategies.NO_FREEZE:
+            get_backbone_for_finetuning = getattr(pl_module, "get_backbone_for_finetuning", None)
+            if get_backbone_for_finetuning is None:
+                raise AttributeError(
+                    "Lightning Module missing instance method 'get_backbone_for_finetuning'."
+                    "Please, implement the method which returns NoneType or a Module or an Iterable of Modules."
+                )
+            modules = get_backbone_for_finetuning()
+            if modules is not None:
+                if isinstance(modules, Module):
+                    modules = [modules]
+                self.freeze(modules=modules, train_bn=self.train_bn)
+
+    def _freeze_unfreeze_function(
         self,
-        pl_module: LightningModule,
+        pl_module: Union[Module, Iterable[Union[Module, Iterable]]],
         epoch: int,
         optimizer: Optimizer,
         opt_idx: int,
-    ) -> None:
-        pass
+        strategy_metadata: int,
+    ):
+        unfreeze_epoch: int = strategy_metadata
+        if epoch != unfreeze_epoch:
+            return
 
+        get_backbone_for_finetuning = getattr(pl_module, "get_backbone_for_finetuning", None)
```

Comment:
IMO, we should avoid the following API. This is quite restrictive and won't scale well to more complex tasks, especially when we start multi modals.

Reply:
I would prefer an advanced scheduler as this one: https://github.com/PyTorchLightning/pytorch-lightning/pull/10198. Which can be used to control the fine-tuning strategy.

Reply:
I understand and agree, this approach might become restrictive for further expansion.

I will currently keep the API simple with just registry based routing of the finetuning callbacks with the added first hook.